src/balancerd/src/lib.rs

```diff
@@ -712,19 +712,17 @@ impl Resolver {
                     _ => anyhow::bail!("expected Password message"),
                 };
 
-                let auth_response = auth
-                    .exchange_password_for_token(&password, user.to_string())
-                    .await;
-                let claims = match auth_response {
-                    Ok(result) => result.claims,
+                let auth_response = auth.authenticate(user.into(), &password).await;
```

Comment:
This looks like the balancer will not reuse the 8-minute auth sessions? Are we ok with that for now?

If invalid creds are presented, do/can we cache those failures in the balancer/envd somehow? Otherwise it'd be trivial to on purpose or accidentally stop re-using auth sessions and start sending lots of requests to frontegg and get rate limited.

I kinda want to teach the balancer to have some LRU cache of (user,password)->tenant, since that won't ever change and envd will still do a full auth session or reuse.

Reply:
> This looks like the balancer will not reuse the 8-minute auth sessions? Are we ok with that for now?

They should be reused during the eight minutes! The session is guaranteed to stick around at least until the initial token expires. It determines whether to clean itself up by checking whether there are still any live receivers _after_ the 8 minutes are up. And yeah, for the balancer layer there is only a tiny window where the receiver is live, so the auth session is almost guaranteed to get cleaned up after that first token expiration. But it will service any requests that arrive within those 8 minutes. Which seemed like a very reasonable place to start!

It's not perfect, but it seemed good enough for now!

> I kinda want to teach the balancer to have some LRU cache of (user,password)->tenant, since that won't ever change and envd will still do a full auth session or reuse.

Yeah, same. I think it'd be as simple as teaching the `Authenticator` to keep track of the last time a session was _used_, and not clean up sessions that were used within the last _n_ minutes. (Rather than just seeing if anyone was listening to the session at the time the token expired.)

> If invalid creds are presented, do/can we cache those failures in the balancer/envd somehow? Otherwise it'd be trivial to on purpose or accidentally stop re-using auth sessions and start sending lots of requests to frontegg and get rate limited.

Not currently cached, but multiple requests that show up with the same invalid credentials at the same time will get merged together.